Sources/AWSSDKSwiftCore/Credential/MetaDataCredentialProvider.swift

```diff
@@ -0,0 +1,258 @@
+            case secretAccessKey = "SecretAccessKey"
+            case token = "Token"
+            case expiration = "Expiration"
+            case roleArn = "RoleArn"
+        }
+    }
+    
+    static let Host = "169.254.170.2"
+    
+    let httpClient    : AWSHTTPClient
+    let endpointURL   : String
+    
+    init?(httpClient: AWSHTTPClient, host: String = ECSMetaDataClient.Host) {
+        guard let relativeURL  = Environment["AWS_CONTAINER_CREDENTIALS_RELATIVE_URI"] else {
+            return nil
+        }
+        
+        self.httpClient     = httpClient
+        self.endpointURL    = "http://\(host)\(relativeURL)"
+    }
+    
+    func getMetaData(on eventLoop: EventLoop) -> EventLoopFuture<ECSMetaData> {
+        return request(url: endpointURL, timeout: 2, on: eventLoop)
+            .flatMapThrowing { response in
+                if let body = response.body {
+                    return try self.decodeResponse(body)
+                }
+                throw MetaDataClientError.failedToDecode
+            }
+    }
+    
+    private func request(url: String, timeout: TimeInterval, on eventLoop: EventLoop) -> EventLoopFuture<AWSHTTPResponse> {
+        let request = AWSHTTPRequest(url: URL(string: url)!, method: .GET, headers: [:], body: nil)
+        return httpClient.execute(request: request, timeout: TimeAmount.seconds(2), on: eventLoop)
+    }
+
+  
+}
+
+//MARK: InstanceMetaDataServiceProvider
+/// Provide AWS credentials for instances
+struct InstanceMetaDataClient: MetaDataClient {
+    typealias MetaData = InstanceMetaData
+
+    struct InstanceMetaData: CredentialContainer {
+        let accessKeyId: String
+        let secretAccessKey: String
+        let token: String
+        let expiration: Date
+        let code: String
+        let lastUpdated: Date
+        let type: String
+
+        var credential: ExpiringCredential {
+            return RotatingCredential(
+                accessKeyId: accessKeyId,
+                secretAccessKey: secretAccessKey,
+                sessionToken: token,
+                expiration: expiration
+            )
+        }
+
+        enum CodingKeys: String, CodingKey {
+            case accessKeyId = "AccessKeyId"
+            case secretAccessKey = "SecretAccessKey"
+            case token = "Token"
+            case expiration = "Expiration"
+            case code = "Code"
+            case lastUpdated = "LastUpdated"
+            case type = "Type"
+        }
+    }
+  
+    static let Host = "169.254.169.254"
+    static let MetadataUri = "/latest/meta-data/iam/security-credentials/"
+
+    let httpClient    : AWSHTTPClient
+    let endpointURL   : String
+  
+    init(httpClient: AWSHTTPClient, host: String = InstanceMetaDataClient.Host, uri: String = InstanceMetaDataClient.MetadataUri) {
+        self.httpClient = httpClient
+        self.endpointURL = "http://\(host)\(uri)"
+    }
+    
+    func getMetaData(on eventLoop: EventLoop) -> EventLoopFuture<InstanceMetaData> {
```

Comment:
You've remove the instance metadata v2 code. Or you based this code off a metadata provider code from prior to it being implemented. We'll need to re-instate it.

Reply:
Can we have a few comments around the instance metadata code. A little comment about each stage would make the code easier to read